### PR TITLE
Added validation and customization function in ghost app

### PR DESCRIPTION
**Pull request by droot**
This PR does the following:

- Adds a package specific validation function for `blog host`
- Adds a customization function for `blog host`
- Misc updates due to running `kpt fn render`

### CODE REVIEW

**Comment by droot** [comment]
@bgrant0607  @yuwenma @justinsb  Added some package specific validation and customization function. Since they are package specific, I wrote them in starlark.

Please take a look and let me know what you think.

**Comment by yuwenma** [comment]
Is that intended to add the labels in line?

File: package-examples/ghost/fn/validate-blog-host.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  def validate_host(resources):
+    errMsg = "Blog's hostname must be specified in fn/update-blog-host.yaml"
+    placeHolderHostname = "example.com"
+    for resource in resources:
+      # this is an abstract package, so ignore host name validation
+      if resource["kind"] == "ConfigMap" and resource["metadata"]["name"] == "kptfile.kpt.dev" and resource["data"]["name"] == "example":
+        return
+    for resource in ctx.resource_list["items"]:
+      if resource["kind"] == "Ingress" and resource["spec"]["rules"][0]["host"] == placeHolderHostname:
+        fail(errMsg)
```
Comment (yuwenma) [inline package-examples/ghost/fn/validate-blog-host.yaml:15]: Sadly, this should be appended to `ResourceList.Results` but it's not quite supported yet.

**Comment by yuwenma** [comment]
I like this approach!

**Review by yuwenma — APPROVED** [review]
Looks great! A minor comment about removing the in-line labels

**Comment by droot** [comment]
> Is that intended to add the labels in line?

Can you elaborate more on this. The labels are added by `set-labels` function and I think  previously published package didn't have the result of `kpt fn render`.

File: package-examples/ghost/fn/update-blog-host.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  for resource in ctx.resource_list["items"]:
+    if resource["kind"] == "Ingress":
+      resource["spec"]["rules"][0]["host"] = host
+
+    if resource["kind"] == "Deployment":
+      container = resource["spec"]["template"]["spec"]["containers"][0]
+
+      for env in container["env"]:
+        if env["name"] == "GHOST_HOST":
+          env["value"] = host
```
Comment (bgrant0607) [inline package-examples/ghost/fn/update-blog-host.yaml:19]: OOC, is ApplyReplacements capable of copying the host value from the Ingress resource to this environment variable in the associative list?

File: package-examples/ghost/fn/validate-blog-host.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+apiVersion: fn.kpt.dev/v1alpha1
+kind: StarlarkRun
+metadata:
+  name: validate-blog-host
+source: |-
+  def validate_host(resources):
+    errMsg = "Blog's hostname must be specified in fn/update-blog-host.yaml"
+    placeHolderHostname = "example.com"
+    for resource in resources:
+      # this is an abstract package, so ignore host name validation
```
Comment (bgrant0607) [inline package-examples/ghost/fn/validate-blog-host.yaml:10]: We probably should have a first-class way to query this in our sdks.

Comment (droot) [reply package-examples/ghost/fn/validate-blog-host.yaml:10]: Yes, validation functions needs this bit, so a helper in SDK will help.

/cc @yuwenma

File: package-examples/ghost/fn/validate-blog-host.yaml (hunk 1 above)
Comment (droot) [reply package-examples/ghost/fn/validate-blog-host.yaml:15]: Ack.

File: package-examples/ghost/fn/update-blog-host.yaml (hunk 2 above)
Comment (droot) [reply package-examples/ghost/fn/update-blog-host.yaml:19]: I will check.

Let's say if ApplyReplacements supports that, then, are you thinking of treating the value of host in `ingress` resource to be source of truth ?

Comment (droot) [reply package-examples/ghost/fn/update-blog-host.yaml:19]: the docs says it is possible to target value in a sequence node, so will give it a try.

https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/replacements/#field-path-format

**Comment by droot** [comment]
Using ApplyReplacements instead of Starlark for propagating the values. Also improved the validation function. So good enough for merging. Also I will be moving the package-examples to kpt-samples repo after tagging this version.

**Pull request merged**